addons/stock/models/stock_picking.py

```diff
@@ -830,6 +830,161 @@ def _create_backorder(self, backorder_moves=[]):
+        with information about move 3.
+
+        :param orig_obj_changes dict: contain a record as key and the
+        change on this record as value.
+        eg: {'move_id': (new product_uom_qty, old product_uom_qty)}
+        :param field_to_group string: It has to be a field of the
+        records that are register in the key of 'orig_obj_changes'
+        eg: 'move_dest_ids' if we use move as record (previous example)
+        :param note_method method: a static function that will generate
+        the html note to log on the activity. The note_method should
+        use the args:
+            - orig_obj_changes
+            - a recordset of a one2many field on which the note is
+            logged. (eg: a list of moves on a picking.)
+            - a dict with link between the orig_obj_changes and the
+            recordset passed as second argument. (eg: If we want to
+            notify the missing quantity from a move to all moves dest
+            we wil pass the moves dest as the recodset but we don't
+            know what are the quantity that changed on the original
+            move. that's the purpose of this dict that will return
+            the original move due to the destination move and we will
+            be able to use the orig_obj_changes in order to find the
+            modified quantity.)
+        the note_method should return a string with an html format
+        :param stream string:
+            - 'UP' if we want to log on the upper most ongoing
+            documents.
+            - 'DOWN' if we want to log on following documents.
+        :param sorted_method method, groupby_method: Only need when
+        stream is 'DOWN', it should sort/group by tuple(object on
+        which the activity is log, the responsible for this object)
+        """
+        orig_obj_child_object_rel = {co: ooc for ooc in orig_obj_changes.keys() for co in ooc[field_to_group]}
+        origin_object = self.env[list(orig_obj_changes.keys())[0]._name].browse([m .id for m in orig_obj_changes.keys()])
+        # The purpose here is to group each destination object by
+        # (document to log, responsible) no matter the stream direction.
+        # example:
+        # {'(delivery_picking_1, admin)': stock.move(1, 2)
+        #  '(delivery_picking_2, admin)': stock.move(3)}
+        if stream == 'DOWN':
+            if sorted_method and groupby_method:
+                child_object_by_parent_object = groupby(sorted(origin_object.mapped(field_to_group), key=sorted_method), key=groupby_method)
+            else:
+                raise UserError(_('You have to define a groupby and sorted method and pass them as arguments.'))
+        elif stream == 'UP':
+            # When using upstream document it is required to define
+            # _get_upstream_documents_and_responsibles on
+            # destination objects in order to ascend documents.
+            child_object_by_parent_object = {}
+            visited_by_parent_object = {}
```

Comment:
visited_documents = {}  # keep a reference to each documents visited to reach the upstream